core/benches/scheduler.rs

```diff
@@ -0,0 +1,494 @@
+        num_works: Arc<AtomicUsize>,
+        num_transaction: Arc<AtomicUsize>,
+        tracer_placement: Arc<AtomicUsize>,
+    ) {
+        while let Ok(work) = work_receiver.recv() {
+            num_works.fetch_add(1, Ordering::Relaxed);
+            let mut tx_count =
+                num_transaction.fetch_add(work.transactions.len(), Ordering::Relaxed);
+            if tracer_placement.load(Ordering::Relaxed) == 0 {
+                work.transactions.iter().for_each(|tx| {
+                    tx_count = tx_count.saturating_add(1);
+                    if is_tracer(tx) {
+                        tracer_placement.store(tx_count, Ordering::Relaxed)
+                    }
+                });
+            }
+
+            if completed_work_sender
+                .send(FinishedConsumeWork {
+                    work,
+                    retryable_indexes: vec![],
+                })
+                .is_err()
+            {
+                // kill this worker if finished_work channel is broken
+                break;
+            }
+        }
+    }
+}
+
+struct BenchEnv<Tx: TransactionWithMeta + Send + Sync + 'static> {
+    #[allow(dead_code)]
+    pingpong_worker: PingPong,
+    filter_1: fn(&[&Tx], &mut [bool]),
+    filter_2: fn(&TransactionState<Tx>) -> PreLockFilterAction,
+    consume_work_senders: Vec<Sender<ConsumeWork<Tx>>>,
+    finished_consume_work_receiver: Receiver<FinishedConsumeWork<Tx>>,
+}
+
+impl<Tx: TransactionWithMeta + Send + Sync + 'static> BenchEnv<Tx> {
+    fn new(stats: &mut BenchStats) -> Self {
+        let num_workers = 4;
+
+        let (consume_work_senders, consume_work_receivers) =
+            (0..num_workers).map(|_| unbounded()).unzip();
+        let (finished_consume_work_sender, finished_consume_work_receiver) = unbounded();
+        let pingpong_worker = PingPong::new(
+            consume_work_receivers,
+            finished_consume_work_sender,
+            stats.num_works.clone(),
+            stats.num_transaction.clone(),
+            stats.tracer_placement.clone(),
+        );
+
+        Self {
+            pingpong_worker,
+            filter_1: Self::test_pre_graph_filter,
+            filter_2: Self::test_pre_lock_filter,
+            consume_work_senders,
+            finished_consume_work_receiver,
+        }
+    }
+
+    fn test_pre_graph_filter(_txs: &[&Tx], results: &mut [bool]) {
+        results.fill(true);
+    }
+
+    fn test_pre_lock_filter(_tx: &TransactionState<Tx>) -> PreLockFilterAction {
+        PreLockFilterAction::AttemptToSchedule
+    }
+
+    fn run(
+        &self,
+        mut scheduler: impl Scheduler<Tx>,
+        mut container: TransactionStateContainer<Tx>,
+        stats: &mut BenchStats,
+    ) {
+        // each bench measurement is to schedule everything in the container
+        while !container.is_empty() {
+            // TODO - should call this, but it panics
+            // scheduler.receive_completed(&mut container).unwrap();
```

Comment:
this panics because the scheduler has been re-constructured each iteration, but the channels to the `PingPong` workers have not been. So the scheduler receives back a batch it never scheduled - which from its logic is an unrecoverable error since something is clearly wrong.

We can take 2 approaches:
1. re-use the same scheduler for each iteration (need to make sure it is completely cleaned).
   - receive finished work for ALL outstanding messages before moving to next iteration
2. create a new schedulder for each iteration. do not send messages back from `PingPong`

Reply:
+1. Was debugging it yesterday, came to same conclusion. I leaning to give each bench a clean setup w new scheduler. was thinking to clean up channels between iterations. But if `PingPong` doesn't send finished_work back, wouldn't that keep txs in container and somewhat changes scheduler perf measure?

Reply:
You start with a fresh container each time, don't you? If you re-use container you need to call `receive_completed`.

Reply:
yes, 10f13ad857e4eb2688401f47815138e787043359 to start with a fresh set of channnls